lib/assertions.js

```diff
@@ -1821,32 +1822,38 @@ module.exports = expect => {
       if (valueType.is('array-like') && !subjectIsArrayLike) {
         expect.fail();
       }
-      const promiseByKey = {};
-      const keys = valueType.getKeys(value);
+
       const subjectKeys = subjectType.getKeys(subject);
+      const valueKeys = valueType.getKeys(value);
+      // calculate the unique keys early given enumerability no
+      // longer affects what is included in the list of keys
+      const uniqueKeys = subjectType.uniqueKeys(subjectKeys, valueKeys);
 
-      if (!subjectIsArrayLike) {
-        // Find all non-enumerable subject keys present in value, but not returned by subjectType.getKeys:
-        keys.forEach(key => {
-          if (
-            Object.prototype.hasOwnProperty.call(subject, key) &&
-            subjectKeys.indexOf(key) === -1
-          ) {
-            subjectKeys.push(key);
+      const promiseByKey = {};
+      let forceExhaustivelyComparison = false;
+      uniqueKeys.forEach((key, index) => {
+        const subjectHasKey = subjectType.hasKey(subject, key);
+        const valueKey = valueType.hasKey(value, key, true)
```

Comment:
I agree that this change is not ideal, but moving the `Object.prototype.hasOwnProperty` check in here would remove the ability for an object-like type to control it. Now that we have all these hook abstracted out, that would be a shame I think. I guess it would affect things like `<Map> to satisfy <Map>` as it wouldn't be possible for the subtype to delegate it to `Map#has`?

Reply:
@papandreou hmm that's a good point. I've been thinking a lot about the semantics so let me reason it out and see if that changes anything.

For something like Map, the keys and values aren't actually stored in any way that affects the prototype chain - whats more, I discovered recently that you can have an own property that conflicts with a key added to the Map via `.set()`. That suggests `<Map> to satisfy <Map>` would use hasKey only for the keys, while `<Map> to satisfy <object>` would be used to check properties attached to the instance. That's why I started to feel the special case should be in `<object> to satisfy <object>` rather than leaking out.

Reply:
Good to be challenged on this - I really don't know what else you can do given the underlying behaviour of the language. This is specifically what I'm thinking of:
```
> m = new Map([])
Map {}
> m.set('foo', 'bar')
Map { 'foo' => 'bar' }
> m.foo = 1
1
> m.get('foo')
'bar'
> m.foo
1
```
In light of the above, I'd argue you'd still need a way to check the actual own properties. If you want to check the contacts of a Map, you use them on the LHS and RHS. And nothing precludes maps backing onto object-like such that, with that type, assertions that need to operate on the maps key/value pairs would be changed to accept `<object-like>`. I'm also suggesting object would be an object-like that happens to serve up its string keys.

Re U-dom, I don't think that analogy holds - we specifically define a DSL, as represented by an normal string keyed object, that allows you to specify the DOMNode. You'd never want to specify those things by passing s Map on the RHS, so it being defined specially as `<object>` is how we avoids breaking things.

Btw, in terms of the DSL, defined something similar for map would  end up with a RHS object of something like:
```js
expect(new Map([]), "to satisfy", {
  keys: [null, undefined, {}],
  values: ['a', 2, null]
});

// or

expect(new Map([]), "to satisfy", [
  { key: null, value: 'a' },
  { key: undefined, value: 2 },
  { key: {}, value: null }
]);
```

Reply:
@papandreou All that side, I mentioned it because I think we can always add the `own` argument later but we'll be hard pressed to remove it after the fact - so until we're sure I'm advocating for being conervative in this case is all :)

Reply:
> Re U-dom, I don't think that analogy holds - we specifically define a DSL, as represented by an normal string keyed object, that allows you to specify the DOMNode. You'd never want to specify those things by passing s Map on the RHS, so it being defined specially as <object> is how we avoids breaking things.

That's not exactly what I meant. My point is that by your own reasoning, a DOM element would also need a way to assert the actual own properties on it, but we're clobbering that by defining `<DOMElement> to satisfy <object>` to mean something else. I think we're too far down that path with `to satisfy` in general to shift gears.

Reply:
Ahh right, I'm with you now - that's an interesting lens on the precedent we have. Will have another think about it in the context of Map semantics.